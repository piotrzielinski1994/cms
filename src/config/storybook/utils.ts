import { ReactNode } from 'react';

const DEFAULT_VALUE = '__default__' as const;
const THUMBNAIL_ID = 'thumbnail';

const storyToUrlPath = (title: string) => {
  const parts = title.split('/');
  const story = parts[parts.length - 1];
  const titlePart = title.toLowerCase().replace(/\//g, '-').replace(/\s+/g, '-');
  const storyPart = story
    .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
    .replace(/([a-zA-Z])(\d+)/g, '$1-$2')
    .toLowerCase();
  return `${titlePart}--${storyPart}`;
};

const getFallback = <T extends ReactNode>(value: T, translation: string): T => {
  return value === DEFAULT_VALUE ? (translation as T) : value;
};

export { DEFAULT_VALUE, getFallback, storyToUrlPath, THUMBNAIL_ID };
