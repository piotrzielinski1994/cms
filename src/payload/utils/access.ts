import { User } from '@/payload.types';
import { Access, FieldAccess } from 'payload';

const anyone: Access = () => true;

const authenticated: Access = ({ req: { user } }) => {
  return Boolean(user);
};

const admin: Access = ({ req: { user } }) => {
  return Boolean(user?.roles?.includes('admin'));
};

const adminOrSelf: Access = ({ req: { user } }) => {
  if (!user) return false;
  if (user.roles?.includes('admin')) return true;
  return { id: { equals: user.id } };
};

const adminFieldLevel: FieldAccess<{ id: string }, User> = ({ req: { user } }) => {
  return Boolean(user?.roles?.includes('admin'));
};

const authenticatedOrPublished: Access = ({ req: { user } }) => {
  if (user) return true;
  return { _status: { equals: 'published' } };
};

export { admin, adminFieldLevel, adminOrSelf, anyone, authenticated, authenticatedOrPublished };
