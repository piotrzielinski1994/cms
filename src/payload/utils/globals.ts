import type { Config } from '@/payload.types';
import configPromise from '@/payload/payload.config';
import { Locale } from 'next-intl';
import { unstable_cache } from 'next/cache';
import { getPayload } from 'payload';

const getGlobal = async (slug: keyof Config['globals'], locale: Locale = 'en', depth = 1) => {
  const payload = await getPayload({ config: configPromise });
  return payload.findGlobal({ slug, depth, locale });
};

const getCachedGlobal = (slug: keyof Config['globals'], locale: Locale, depth = 1) => {
  const cacheConfig = { tags: [`global__${locale}_${slug}`] };
  return unstable_cache(() => getGlobal(slug, locale, depth), [slug, locale], cacheConfig);
};

export { getCachedGlobal };
