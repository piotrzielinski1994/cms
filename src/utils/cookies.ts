import { setCookie } from 'typescript-cookie';

const setPermament = (name: string, value: string | number | boolean | object) => {
  setCookie(name, value.toString(), { expires: 365 * 10, path: '/' });
};

const cookies = { setPermament };

export default cookies;
