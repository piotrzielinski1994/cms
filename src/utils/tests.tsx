import { Providers } from '@/providers';
import { ComponentProps, ReactNode } from 'react';

const resolveServerComponent = async <T extends (props: unknown) => Promise<ReactNode>>(
  Component: T,
  props: Parameters<T>[0],
) => {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
};

const withProviders = (
  children: ReactNode,
  props: Partial<Omit<ComponentProps<typeof Providers>, 'children'>> = {},
) => {
  const overwrittenProps: ComponentProps<typeof Providers> = {
    locale: 'en',
    initialTheme: 'light',
    initialFontScale: 'base',
    initialCookiesConsent: true,
    ...props,
    children,
  };
  return <Providers {...overwrittenProps} />;
};

export { resolveServerComponent, withProviders };
