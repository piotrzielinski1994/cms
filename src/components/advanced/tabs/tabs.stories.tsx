import { DEFAULT_VALUE, getFallback } from '@/config/storybook/utils';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { type ComponentProps } from 'react';
import { Tabs as TabsComponent } from './tabs';

type Args = ComponentProps<typeof TabsComponent>;

const meta = {
  component: TabsComponent,
  title: 'Advanced/Tabs',
  argTypes: {
    tabs: { control: 'object' },
  },
  args: {
    tabs: [
      {
        heading: DEFAULT_VALUE,
        content: DEFAULT_VALUE,
      },
      {
        heading: DEFAULT_VALUE,
        content: DEFAULT_VALUE,
      },
    ],
  },
} satisfies Meta<Args>;

const Render = ({ tabs }: Args) => {
  const t = useTranslations('storybook.basic.tabs');
  const props = {
    tabs: tabs.map((tab, index) => ({
      heading: getFallback(tab.heading, `${t('tab')} ${index + 1}`),
      content: getFallback(tab.content, `${t('content')} ${index + 1}`),
    })),
  } satisfies Args;

  return <TabsComponent {...props} />;
};

const Tabs: StoryObj<typeof TabsComponent> = { render: Render };

export { Tabs };
export default meta;
