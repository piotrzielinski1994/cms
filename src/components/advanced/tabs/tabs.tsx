import { useHtmlId } from '@/utils/html/html.hooks';
import { cn } from '@/utils/tailwind';
import { ReactNode, useState } from 'react';

type TabsProps = {
  tabs: Array<{
    heading: ReactNode;
    content: ReactNode;
  }>;
};

const styles = {
  root: 'grid',
  header: 'flex gap-2',
  tab: cn(
    'has-[:checked]:border-b border-foreground',
    'tw-has-focus:tw-cms-outline',
    'px-4 py-2 md:px-6',
    'cursor-pointer',
  ),
  content: 'p-4 md:px-6',
  radio: 'peer sr-only',
};

const Tabs = ({ tabs }: TabsProps) => {
  const [activeIndex, setActiveIndex] = useState(0);
  const { id, getId } = useHtmlId('tabs');

  return (
    <div className={styles.root}>
      <div className={styles.header}>
        {tabs.map((tab, index) => {
          return (
            <label
              key={index}
              id={getId('tab', index)}
              aria-controls={getId('panel', index)}
              className={styles.tab}
            >
              <input
                type="radio"
                name={`tabs-${id}`}
                className={styles.radio}
                checked={index === activeIndex}
                onChange={() => setActiveIndex(index)}
              />
              {tab.heading}
            </label>
          );
        })}
      </div>
      {tabs.map((tab, index) => (
        <div
          key={index}
          role="tabpanel"
          id={getId('panel', index)}
          aria-labelledby={getId('tab', index)}
          hidden={index !== activeIndex}
          className={styles.content}
        >
          {tab.content}
        </div>
      ))}
    </div>
  );
};

export { Tabs };
