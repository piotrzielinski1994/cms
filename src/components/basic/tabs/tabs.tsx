import { cn } from '@/utils/tailwind';
import { useTranslations } from 'next-intl';
import { ReactNode, useId, useState } from 'react';

type TabsProps = {
  tabs: Array<{
    heading: ReactNode;
    content: ReactNode;
  }>;
};

const Tabs = ({ tabs }: TabsProps) => {
  const t = useTranslations('frontend.component');
  const [activeIndex, setActiveIndex] = useState(0);
  const id = useId();

  return (
    <div className="grid gap-2">
      <div aria-label={t('tabs')} className="flex gap-2">
        {tabs.map((tab, index) => (
          <label
            key={index}
            id={tabId(id, index)}
            aria-controls={panelId(id, index)}
            className={cn(
              'has-[:checked]:border-b border-foreground',
              '[&:has(input:focus-visible)]:tw-cms-outline',
              'px-2',
              'cursor-pointer',
            )}
          >
            <input
              type="radio"
              name={`tabs-${id}`}
              className="peer sr-only"
              checked={index === activeIndex}
              onChange={() => setActiveIndex(index)}
            />
            {tab.heading}
          </label>
        ))}
      </div>
      {tabs.map((tab, index) => (
        <div
          key={index}
          role="tabpanel"
          id={panelId(id, index)}
          aria-labelledby={tabId(id, index)}
          hidden={index !== activeIndex}
        >
          {tab.content}
        </div>
      ))}
    </div>
  );
};

const tabId = (id: string, index: number) => `tabs__${id}__${index}`;
const panelId = (id: string, index: number) => `tabs__${id}__${index}`;

export { Tabs };
