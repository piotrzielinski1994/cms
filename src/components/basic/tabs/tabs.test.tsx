import { withProviders } from '@/utils/tests';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ComponentProps } from 'react';
import { describe, expect, it } from 'vitest';
import { axe } from 'vitest-axe';
import { Tabs } from './tabs';

describe('Tabs', () => {
  const defaultProps = {
    tabs: [
      {
        heading: 'Tab 1',
        content: 'Content 1',
      },
      {
        heading: 'Tab 2',
        content: 'Content 2',
      },
    ],
  } satisfies ComponentProps<typeof Tabs>;

  it('should have no accessibility violations', async () => {
    const { container } = render(withProviders(<Tabs {...defaultProps} />));
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(withProviders(<Tabs {...defaultProps} />));
    expect(container).toMatchSnapshot();
  });

  it('should show content of the first tab initially', () => {
    const { queryByText } = render(withProviders(<Tabs {...defaultProps} />));
    expect(queryByText('Content 1')).toBeVisible();
    expect(queryByText('Content 2')).not.toBeVisible();
  });

  it('should switch content when tab is clicked', async () => {
    const { getByText, queryByText } = render(withProviders(<Tabs {...defaultProps} />));
    await userEvent.click(getByText('Tab 2'));
    expect(queryByText('Content 2')).toBeVisible();
    expect(queryByText('Content 1')).not.toBeVisible();
  });
});
