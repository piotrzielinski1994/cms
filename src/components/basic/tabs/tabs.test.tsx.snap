// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs > should match the snapshot 1`] = `
<div>
  <div
    class="grid gap-2"
  >
    <div
      aria-label="Tabs"
      class="flex gap-2"
    >
      <label
        aria-controls="tabs__«r1»__0"
        class="has-[:checked]:border-b border-foreground [&:has(input:focus-visible)]:tw-cms-outline px-2 cursor-pointer"
        id="tabs__«r1»__0"
      >
        <input
          checked=""
          class="peer sr-only"
          name="tabs-«r1»"
          type="radio"
        />
        Tab 1
      </label>
      <label
        aria-controls="tabs__«r1»__1"
        class="has-[:checked]:border-b border-foreground [&:has(input:focus-visible)]:tw-cms-outline px-2 cursor-pointer"
        id="tabs__«r1»__1"
      >
        <input
          class="peer sr-only"
          name="tabs-«r1»"
          type="radio"
        />
        Tab 2
      </label>
    </div>
    <div
      aria-labelledby="tabs__«r1»__0"
      id="tabs__«r1»__0"
      role="tabpanel"
    >
      Content 1
    </div>
    <div
      aria-labelledby="tabs__«r1»__1"
      hidden=""
      id="tabs__«r1»__1"
      role="tabpanel"
    >
      Content 2
    </div>
  </div>
</div>
`;
